Added docs for API denylists (#3489)
dapr/docs: daprdocs/content/en/operations/configuration/api-allowlist.md

description: "Choose which Dapr sidecar APIs are available to the app"
---

In certain scenarios, such as zero trust networks or when exposing the Dapr sidecar to external traffic through a frontend, it's recommended to only enable the Dapr sidecar APIs that are being used by the app. Doing so reduces the attack surface and helps keep the Dapr APIs scoped to the actual needs of the application.

Dapr allows developers to control which APIs are accessible to the application by setting an API allowlist or denylist using a [Dapr Configuration]({{<ref "configuration-overview.md">}}).

### Default behavior

If no API allowlist or denylist is specified, the default behavior is to allow access to all Dapr APIs.

- If only a denylist is defined, all Dapr APIs are allowed except those defined in the denylist
- If only an allowlist is defined, only the Dapr APIs listed in the allowlist are allowed

> - If both an allowlist and a denylist are defined, the allowed APIs are those defined in the allowlist, unless they are also included in the denylist. In other terms, the denylist overrides the allowlist for APIs that are defined in both places.

Reviewer: ```suggestion
- If both an allowlist and a denylist are defined, the allowed APIs are those defined in the allowlist, unless they are also included in the denylist. In other words, the denylist overrides the allowlist for APIs that are defined in both.
```

    samplingRate: "1"
```

### Using an allowlist

#### Enabling specific HTTP APIs

> The following example enables the state `v1.0` HTTP API and block all other HTTP APIs:

Reviewer: ```suggestion
The following example enables the state `v1.0` HTTP API and blocks all other HTTP APIs:
```

spec:
  api:
    allowed:
      - name: state
        version: v1.0
        protocol: http
```

#### Enabling specific gRPC APIs

> The following example enables the state `v1` gRPC API and block all other gRPC APIs:

Reviewer: ```suggestion
The following example enables the state `v1` gRPC API and blocks all other gRPC APIs:
```

spec:
  api:
    allowed:
      - name: state
        version: v1
        protocol: grpc
```

### Using a denylist

#### Disabling specific HTTP APIs

The following example disables the state `v1.0` HTTP API, allowing all other HTTP APIs:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: myappconfig
  namespace: default
spec:
  api:
    denied:
      - name: state
        version: v1.0
        protocol: http
```

#### Disabling specific gRPC APIs

The following example disables the state `v1` gRPC API, allowing all other gRPC APIs:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: myappconfig
  namespace: default
spec:
  api:
    denied:
      - name: state
        version: v1
        protocol: grpc
```

> ### List of Dapr APIs

Reviewer: Just seen that this list does not include "Configuation" which is now V1 and should be enabled.

Also I presume that **all** alpha API can, like workflow, Lock and Crypto can also be blocked. If they can they provide an example of what is needed for the version. If not then, we should understand why and put a comment here saying the reason for not blocking them, since this may be a security concern.

Author: Yes, the list below should probably be revised. Let's do that in a separate PR.

Yup, all APIs can be blocked, even alpha ones.

Reviewer: Why not add to this PR?

Reviewer: @ItalyPaleAle - Need you to confirm the API values to ensure this list is correct. Can you please take look?

Author: Please take a look now

See this list of values corresponding to the different Dapr APIs:

| API group | HTTP API | [gRPC API](https://github.com/dapr/dapr/blob/master/pkg/grpc/endpoints.go) |
| ----- | ----- | ----- |
| [Service Invocation]({{< ref service_invocation_api.md >}}) | `invoke` (`v1.0`) | `invoke` (`v1`) |

> | [State]({{< ref state_api.md>}})| `state` (`v1.0` and `v1.0-alpha1`) | `state` (`v1` and `v1alpha1`) |

Reviewer: Are these "or" rather than "and". You can use either entry, yes? You can either do v1.0 or v1.0-alpha.
```suggestion
| [State]({{< ref state_api.md>}})| `state` (`v1.0` or `v1.0-alpha1`) | `state` (`v1` or `v1alpha1`) |
```

Author: It's an "and". You can use both (depending on your needs)

> | [Pub/Sub]({{< ref pubsub.md >}}) | `publish` (`v1.0` and `v1.0-alpha1`) | `publish` (`v1` and `v1alpha1`) |

Reviewer: Or?

Author: Ditto

| [(Output) Bindings]({{< ref bindings_api.md >}})  | `bindings` (`v1.0`) |`bindings` (`v1`) |
| [Secrets]({{< ref secrets_api.md >}})| `secrets` (`v1.0`) | `secrets` (`v1`) |
| [Actors]({{< ref actors_api.md >}}) | `actors`  (`v1.0`) |`actors` (`v1`) |
| [Metadata]({{< ref metadata_api.md >}}) | `metadata` (`v1.0`) |`metadata` (`v1`) |

> | [Configuration]({{< ref configuration_api.md >}}) | `configuration` (`v1.0` and `v1.0-alpha1`) | `configuration` (`v1` and `v1alpha1`) |

Reviewer: Really just want to encourage V1 
```suggestion
| [Configuration]({{< ref configuration_api.md >}}) | `configuration` (`v1.0`) | `configuration` (`v1`) |
```

Author: It's not about encouraging anything... If you maintain a denylist and only include `v1.0`, then `v1.0-alpha1` can still be used. I've listed all options there.
